sno/structure.py

```diff
@@ -602,11 +599,152 @@ def diff_meta(self, other, reverse=False):
         meta_new = dict(new.iter_meta_items()) if new else {}
         return DeltaDiff.diff_dicts(meta_old, meta_new)
 
-    def write_index(self, dataset_diff, index, repo):
+    def _replace_subtree(self, repo, root_tree, subpath, subtree_or_blob):
```

Comment:
Move at least _replace_subtree into a new file:
tree_util.py

_recursive_build_feature_tree is a bit of a hybrid - it has some code that is feature-aware, but also some code for dealing with trees. Ideally we would split this into two halves, the feature part that can stay here and the tree part that would also go to tree_util, and we would figure out what sort of diff to send to tree_util to give it the info it needs. Something like a DeltaDiff, but lower-level - I guess something more like a pygit2.Diff although that probably is a horrible type with horrible performance.

For now you can just move _replace_subtree.